pkg/services/ngalert/state/state_tracker.go

```diff
@@ -53,22 +54,36 @@ func NewStateTracker(logger log.Logger) *StateTracker {
 	return tracker
 }
 
-func (st *StateTracker) getOrCreate(uid string, orgId int64, result eval.Result) AlertState {
+func (st *StateTracker) getOrCreate(alertRule *ngModels.AlertRule, result eval.Result) AlertState {
 	st.stateCache.mu.Lock()
 	defer st.stateCache.mu.Unlock()
+	lbs := alertRule.Labels
+	if len(result.Instance) > 0 {
+		lbs = result.Instance
+	}
+	lbs["__alert_rule_uid__"] = alertRule.UID
+	lbs["__alert_rule_namespace_uid__"] = alertRule.NamespaceUID
+	lbs["__alert_rule_title__"] = alertRule.Title
```

Comment:
(Assuming this will be `alertname` instead) Currently the notifier assumes the labels wont have any blank spaces, and looking at some example rules till now, I see title can be anything, even with spaces. This can cause issues while rendering the alert.  Do we want to enforce any pattern for the title? Or remove the whitespaces from title and camel case it for label?

Reply:
I'd say we want both Ganesh, the auto-transformation serves its purpose for the migration and then validate it at an API / reception level.

Reply:
@codesome I can work with that if we need to. Can you and @gotjosh decide on the constraints and let me know? I'm still not sure exactly what the set of default labels should be.